compiler/passes/src/storage_lowering/ast.rs

```diff
@@ -0,0 +1,969 @@
+                statements.extend(stmts);
+                expr
+            })
+            .collect();
+
+        (Type::Composite(CompositeType { const_arguments, ..input }), statements)
+    }
+
+    /* Expressions */
+    fn reconstruct_array_access(
+        &mut self,
+        mut input: ArrayAccess,
+        _additional: &(),
+    ) -> (Expression, Self::AdditionalOutput) {
+        let (array, mut stmts_array) = self.reconstruct_expression(input.array, &());
+        let (index, mut stmts_index) = self.reconstruct_expression(input.index, &());
+
+        input.array = array;
+        input.index = index;
+
+        // Merge side effects
+        stmts_array.append(&mut stmts_index);
+
+        (input.into(), stmts_array)
+    }
+
+    fn reconstruct_associated_function(
+        &mut self,
+        mut input: AssociatedFunctionExpression,
+        _additional: &(),
+    ) -> (Expression, Self::AdditionalOutput) {
+        match CoreFunction::from_symbols(input.variant.name, input.name.name) {
+            Some(CoreFunction::VectorPush) => {
+                // Input:
+                //   Vector::push(v, 42u32)
+                //
+                // Lowered reconstruction:
+                //   let $len_var = Mapping::get_or_use(len_map, false, 0u32);
+                //   Mapping::set(vec_map, $len_var, 42u32);
+                //   Mapping::set(len_map, false, $len_var + 1);
+
+                // Unpack arguments
+                let [vector_expr, value_expr] = &mut input.arguments[..] else {
+                    panic!("Vector::push should have 2 arguments");
+                };
+
+                // Validate vector type
+                assert!(matches!(self.state.type_table.get(&vector_expr.id()), Some(Type::Vector(_))));
+
+                // Reconstruct value
+                let (value, stmts) = self.reconstruct_expression(value_expr.clone(), &());
+
+                let (vec_values_mapping_name, vec_length_mapping_name) =
+                    self.generate_mapping_names_for_vector(vector_expr);
+                let vec_path_expr = self.symbol_to_path_expr(vec_values_mapping_name);
+                let len_path_expr = self.symbol_to_path_expr(vec_length_mapping_name);
+
+                // let $len_var = Mapping::get_or_use(len_map, false, 0u32)
+                let len_var_sym = self.state.assigner.unique_symbol("$len_var", "$");
+                let len_var_ident =
+                    Identifier { name: len_var_sym, span: Default::default(), id: self.state.node_builder.next_id() };
+                let get_len_expr = self.get_vector_len_expr(len_path_expr.clone(), input.span);
+                let len_stmt = self.state.assigner.simple_definition(
+                    len_var_ident,
+                    get_len_expr,
+                    self.state.node_builder.next_id(),
+                );
+                let len_var_expr: Expression = len_var_ident.into();
```

Comment:
In a lot of our lowering passes, we create new AST nodes and invoke `self.state.node_builder.next_id()`. Should type information be propagated to these nodes?

Reply:
For this pass and many other early ones, we run the type checker immediately after they're done. We actually do that in the pass's `mod.rs` directly so that the next pass that runs after `StorageLowering` has all the necessary type information.